File: tests/db/support/mysql.ts
import { resolve } from 'node:path';
import type { ResolvedConnectionConfig } from '@shared/protocol/engine-ops';
import { MySqlContainer, type StartedMySqlContainer } from '@testcontainers/mysql';
import { createConnection, importFile } from 'mariadb';
import { Wait } from 'testcontainers';
import { resolveDockerHost } from './docker';

resolveDockerHost();

// P34 D24: MySQL's current LTS — defaults to caching_sha2_password, has dropped
// --default-authentication-plugin in favour of authentication_policy, and ships
// mysql_native_password present-but-disabled. This is the version D3/D4's auth handling exists for.
const IMAGE = 'mysql:8.4';
const ROOT_PASSWORD = 'kira';
const DATABASE = 'kira_test';
const ANALYTICS_DATABASE = 'kira_analytics';
const USERNAME = 'kira';
const PASSWORD = 'kira';
const STARTUP_TIMEOUT_MS = 120_000;
const SEED_SQL_PATH = resolve(__dirname, '../fixtures/0008_mysql_seed.sql');

// P34 D29/F4: never-authenticated users, created once by the fixture solely for the
// plaintext-caching_sha2_password scenario. After any successful authentication, MySQL's SHA2
// cache serves the fast path and a plaintext connection for that user *succeeds* — so the scenario
// needs a user whose first-ever authentication is the one under test. Two, not one: scenario 2b's
// (a)+(b) subparts reuse NOCACHE_USER (the (a) attempt fails closed, so it never warms the cache;
// the (b) attempt succeeds over TLS and does warm it, but nothing plaintext against this user runs
// after that), and (c) needs its own still-cold user, PUBKEY_USER.
export const NOCACHE_USER = 'kira_nocache';
export const PUBKEY_USER = 'kira_pubkey';
const NOCACHE_PASSWORD = 'kira';

export interface MysqlFixture {
  container: StartedMySqlContainer;
  config: ResolvedConnectionConfig; // ready to hand to the adapter
  uri: string;
  stop(): Promise<void>;
}

/** A connection config for one of the never-authenticated users above, options merged onto the
 *  fixture's own host/port/database — never the fixture's own `config` (that carries `sslmode:
 *  'require'` and the main `kira` user, D26). */
export function nocacheConfig(
  fixture: MysqlFixture,
  username: string,
  options: Record<string, string>,
): ResolvedConnectionConfig {
  return { ...fixture.config, username, password: NOCACHE_PASSWORD, options };
}

// One container per test process, same discipline as support/mariadb.ts (§11b).
let memoized: Promise<MysqlFixture> | null = null;

export function startMysql(opts?: { seedBigTable?: boolean }): Promise<MysqlFixture> {
  if (!memoized) memoized = start(opts);
  return memoized;
}

async function start(opts?: { seedBigTable?: boolean }): Promise<MysqlFixture> {
  // P34 D25/F25: @testcontainers/mysql sets no wait strategy or healthcheck of its own (as of
  // 12.1.0, same gap support/mariadb.ts:36-38 documents for @testcontainers/mariadb) — and the
  // official MySQL entrypoint boots the server twice during initialisation, the first boot on a
  // socket only, so a TCP-reaching healthcheck is what actually distinguishes "initialising" from
  // "ready" (the same double-boot hazard the MariaDB fixture's own comment names).
  const container = await new MySqlContainer(IMAGE)
    .withDatabase(DATABASE)
    .withUsername(USERNAME)
    .withUserPassword(PASSWORD)
    .withRootPassword(ROOT_PASSWORD)
    .withHealthCheck({
      test: [
        'CMD',
        'mysqladmin',
        'ping',
        '-h',
        '127.0.0.1',
        '-u',
        'root',
        `-p${ROOT_PASSWORD}`,
        '--silent',
      ],
      interval: 2000,
      timeout: 5000,
      retries: 60,
    })
    .withWaitStrategy(Wait.forHealthCheck())
    .withStartupTimeout(STARTUP_TIMEOUT_MS)
    .start();

  const host = container.getHost();
  const port = container.getPort();
  const uri = `mysql://${USERNAME}:${PASSWORD}@${host}:${port}/${DATABASE}`;

  // Seeded as root (CREATE/GRANT need it); the returned config connects as the unprivileged
  // `kira` user — the entrypoint already grants it full rights on kira_test, so the read path
  // proves KILL QUERY on your own query needs no PROCESS privilege (§6e), same as MariaDB's.
  await importFile({
    host,
    port,
    user: 'root',
    password: ROOT_PASSWORD,
    database: DATABASE,
    multipleStatements: true,
    ssl: false,
    file: SEED_SQL_PATH,
  });

  const rootConn = await createConnection({
    host,
    port,
    user: 'root',
    password: ROOT_PASSWORD,
    database: DATABASE,
    ssl: false,
  });
  try {
    // A second database, mirroring the MariaDB fixture's kira_analytics (support/mariadb.ts:91-93)
    // — MySQL has no schema level either, so a second *database* is the equivalent (§6d).
    await rootConn.query(`CREATE DATABASE IF NOT EXISTS \`${ANALYTICS_DATABASE}\``);
    await rootConn.query(
      `CREATE TABLE IF NOT EXISTS \`${ANALYTICS_DATABASE}\`.events (
         id INT AUTO_INCREMENT PRIMARY KEY,
         event_name VARCHAR(255) NOT NULL,
         occurred_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
       )`,
    );
    await rootConn.query(
      `INSERT INTO \`${ANALYTICS_DATABASE}\`.events (event_name) VALUES ('signup'), ('login')`,
    );
    await rootConn.query(`GRANT SELECT ON \`${ANALYTICS_DATABASE}\`.* TO '${USERNAME}'@'%'`);

    // P34 D29: created here, as a plain CREATE USER over the root connection — that authenticates
    // *root*, never these two users themselves, so their own SHA2 cache stays cold until a
    // scenario deliberately authenticates as one of them.
    await rootConn.query(
      `CREATE USER IF NOT EXISTS '${NOCACHE_USER}'@'%' IDENTIFIED BY '${NOCACHE_PASSWORD}'`,
    );
    await rootConn.query(`GRANT SELECT ON \`${DATABASE}\`.* TO '${NOCACHE_USER}'@'%'`);
    await rootConn.query(
      `CREATE USER IF NOT EXISTS '${PUBKEY_USER}'@'%' IDENTIFIED BY '${NOCACHE_PASSWORD}'`,
    );
    await rootConn.query(`GRANT SELECT ON \`${DATABASE}\`.* TO '${PUBKEY_USER}'@'%'`);
    await rootConn.query('FLUSH PRIVILEGES');

    // P34 D28: MySQL has no SEQUENCE storage engine (F13), so the 1,000,000-row bulk insert uses
    // the conventional numbers-table idiom instead — a six-way cross join over a 10-row digits
    // table, chunked into ten 100,000-row statements (one per outer-digit value) so any single
    // InnoDB transaction stays bounded. Materialisation-free, unlike a 1,000,000-deep recursive
    // CTE, which would need cte_max_recursion_depth raised past its 1,000 default.
    if (opts?.seedBigTable ?? true) {
      await rootConn.query('CREATE TEMPORARY TABLE digits (d INT PRIMARY KEY)');
      await rootConn.query('INSERT INTO digits (d) VALUES (0),(1),(2),(3),(4),(5),(6),(7),(8),(9)');
      for (let outer = 0; outer < 10; outer++) {
        await rootConn.query(
          `INSERT INTO \`${DATABASE}\`.big_rows (id, payload)
           SELECT d1.d*100000 + d2.d*10000 + d3.d*1000 + d4.d*100 + d5.d*10 + d6.d + 1 AS id,
                  MD5(d1.d*100000 + d2.d*10000 + d3.d*1000 + d4.d*100 + d5.d*10 + d6.d + 1) AS payload
           FROM digits d1, digits d2, digits d3, digits d4, digits d5, digits d6
           WHERE d1.d = ?`,
          [outer],
        );
      }
      await rootConn.query(`ANALYZE TABLE \`${DATABASE}\`.big_rows`);
    }
  } finally {
    await rootConn.end();
  }

  const now = new Date().toISOString();
  const config: ResolvedConnectionConfig = {
    id: 'test-mysql',
    sortOrder: 0,
    createdAt: now,
    updatedAt: now,
    name: 'Test MySQL',
    kind: 'mysql',
    color: 'teal',
    mode: 'fields',
    readOnly: false,
    host,
    port,
    database: DATABASE,
    username: USERNAME,
    uri: null,
    // P34 D5/D26: the documented working configuration against a stock MySQL 8 server — TLS is
    // available on a stock server (MySQL auto-generates a self-signed certificate at init), so
    // this exercises the real remedy path rather than working by accident (F5).
    options: { sslmode: 'require' },
    password: PASSWORD,
  };

  return {
    container,
    config,
    uri,
    async stop() {
      // Playwright's workers:1 config runs every UI spec file sequentially in the same worker
      // process, sharing this module's state — without resetting `memoized`, a later spec file's
      // startMysql() would return this now-dead container instead of starting a fresh one.
      memoized = null;
      await container.stop();
    },
  };
}

File: shell/internal/adapters/testsupport/mysql.go
package testsupport

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"testing"
	"time"

	sqldriver "github.com/go-sql-driver/mysql"
	"github.com/testcontainers/testcontainers-go"
	tcmysql "github.com/testcontainers/testcontainers-go/modules/mysql"

	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// MysqlFixture is support/mysql.ts's MysqlFixture.
type MysqlFixture struct {
	URI       string
	Config    model.ResolvedConnectionConfig
	container testcontainers.Container
}

var mysqlMemo fixture[MysqlFixture]

const (
	mysqlImage        = "mysql:8.4"
	mysqlRootPassword = "kira"
	mysqlDatabase     = "kira_test"
	mysqlAnalyticsDB  = "kira_analytics"
	mysqlUsername     = "kira"
	mysqlPassword     = "kira"
)

// StartMysql is support/mysql.ts's startMysql. See fixture.go's own doc comment for why
// termination is never wired to t.Cleanup (B15).
func StartMysql(t *testing.T) *MysqlFixture {
	t.Helper()
	if !IsDockerAvailable() {
		t.Skip(DockerUnavailableMessage)
	}
	fixture, err := mysqlMemo.get(startMysql)
	if err != nil {
		t.Fatalf("mysql container: %v", err)
	}
	return fixture
}

// StopMysql terminates the memoized container, if one was ever started. Call once, from the test
// binary's own TestMain, after m.Run() returns — never from an individual test.
func StopMysql() {
	mysqlMemo.stop(func(f *MysqlFixture) { _ = f.container.Terminate(context.Background()) })
}

func startMysql() (*MysqlFixture, error) {
	ctx := context.Background()

	container, err := tcmysql.Run(ctx, mysqlImage,
		tcmysql.WithDatabase(mysqlDatabase),
		tcmysql.WithUsername(mysqlUsername),
		tcmysql.WithPassword(mysqlPassword),
		testcontainers.WithEnv(map[string]string{"MYSQL_ROOT_PASSWORD": mysqlRootPassword}),
	)
	if err != nil {
		return nil, err
	}

	host, err := container.Host(ctx)
	if err != nil {
		return nil, err
	}
	port, err := container.MappedPort(ctx, "3306/tcp")
	if err != nil {
		return nil, err
	}
	portNum := int(port.Num())

	seedPath := filepath.Join(repoRoot(), "tests", "db", "fixtures", "0008_mysql_seed.sql")
	seedSQL, err := os.ReadFile(seedPath)
	if err != nil {
		return nil, err
	}

	rootDSN := rootMysqlDSN(host, portNum, mysqlDatabase)
	if err := seedMysqlFamilyDatabase(ctx, rootDSN, string(seedSQL)); err != nil {
		return nil, err
	}
	if err := seedMysqlExtras(ctx, host, portNum); err != nil {
		return nil, err
	}

	now := time.Now().UTC().Format(time.RFC3339Nano)
	hostCopy, portCopy := host, portNum
	cfg := model.ResolvedConnectionConfig{
		ID: "test-mysql", SortOrder: 0, CreatedAt: now, UpdatedAt: now,
		Name: "Test MySQL", Kind: "mysql", Color: "teal", Mode: "fields", ReadOnly: false,
		Host: &hostCopy, Port: &portCopy, Database: Strp(mysqlDatabase), Username: Strp(mysqlUsername),
		// P34 D5/D26: TLS is available on a stock server (MySQL auto-generates a self-signed
		// certificate at init) — this exercises the real remedy path, not a worked-around one.
		Options:  map[string]any{"sslmode": "require"},
		Password: Strp(mysqlPassword),
	}
	uri := fmt.Sprintf("mysql://%s:%s@%s:%d/%s", mysqlUsername, mysqlPassword, host, portNum, mysqlDatabase)
	return &MysqlFixture{URI: uri, Config: cfg, container: container}, nil
}

func rootMysqlDSN(host string, port int, database string) string {
	cfg := sqldriver.NewConfig()
	cfg.Net = "tcp"
	cfg.Addr = fmt.Sprintf("%s:%d", host, port)
	cfg.User = "root"
	cfg.Passwd = mysqlRootPassword
	cfg.DBName = database
	cfg.MultiStatements = true
	cfg.AllowNativePasswords = true
	return cfg.FormatDSN()
}

// seedMysqlExtras is support/mysql.ts's own root-connection block: the second database (mirroring
// MariaDB's kira_analytics, §6d) and the big_rows bulk insert. P34 D28: MySQL has no SEQUENCE
// storage engine, so the 1,000,000-row insert uses the conventional numbers-table idiom instead —
// a six-way cross join over a 10-row digits table, chunked into ten 100,000-row statements so any
// single InnoDB transaction stays bounded.
func seedMysqlExtras(ctx context.Context, host string, port int) error {
	db, err := sql.Open("mysql", rootMysqlDSN(host, port, mysqlDatabase))
	if err != nil {
		return err
	}
	defer db.Close()

	if _, err := db.ExecContext(ctx, "CREATE DATABASE IF NOT EXISTS `"+mysqlAnalyticsDB+"`"); err != nil {
		return err
	}
	if _, err := db.ExecContext(ctx, "CREATE TABLE IF NOT EXISTS `"+mysqlAnalyticsDB+"`.events ("+
		"id INT AUTO_INCREMENT PRIMARY KEY, event_name VARCHAR(255) NOT NULL, "+
		"occurred_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP)"); err != nil {
		return err
	}
	if _, err := db.ExecContext(ctx, "INSERT INTO `"+mysqlAnalyticsDB+"`.events (event_name) VALUES ('signup'), ('login')"); err != nil {
		return err
	}
	if _, err := db.ExecContext(ctx, "GRANT SELECT ON `"+mysqlAnalyticsDB+"`.* TO '"+mysqlUsername+"'@'%'"); err != nil {
		return err
	}
	if _, err := db.ExecContext(ctx, "FLUSH PRIVILEGES"); err != nil {
		return err
	}

	// A plain table, not TEMPORARY: MySQL's TEMPORARY tables can't be referenced more than once in
	// the same query, and the six-way self-join below does exactly that. Dropped once done.
	if _, err := db.ExecContext(ctx, "CREATE TABLE digits (d INT PRIMARY KEY)"); err != nil {
		return err
	}
	if _, err := db.ExecContext(ctx, "INSERT INTO digits (d) VALUES (0),(1),(2),(3),(4),(5),(6),(7),(8),(9)"); err != nil {
		return err
	}
	for outer := 0; outer < 10; outer++ {
		if _, err := db.ExecContext(ctx,
			fmt.Sprintf(`INSERT INTO `+"`%s`"+`.big_rows (id, payload)
			 SELECT d1.d*100000 + d2.d*10000 + d3.d*1000 + d4.d*100 + d5.d*10 + d6.d + 1 AS id,
			        MD5(d1.d*100000 + d2.d*10000 + d3.d*1000 + d4.d*100 + d5.d*10 + d6.d + 1) AS payload
			 FROM digits d1, digits d2, digits d3, digits d4, digits d5, digits d6
			 WHERE d1.d = ?`, mysqlDatabase), outer,
		); err != nil {
			return err
		}
	}
	if _, err := db.ExecContext(ctx, "DROP TABLE digits"); err != nil {
		return err
	}
	if _, err := db.ExecContext(ctx, "ANALYZE TABLE `"+mysqlDatabase+"`.big_rows"); err != nil {
		return err
	}
	return nil
}
